packages/cluster-ui/src/store/terminateQuery/terminateQuery.sagas.ts

```diff
@@ -11,49 +11,74 @@
 import { PayloadAction } from "@reduxjs/toolkit";
 import { all, call, put, takeEvery } from "redux-saga/effects";
 
-// TODO: (vlad) originaly this calls triggered popup message with error\success message.
-// need to come up with api to recreate this behavier with client specific popup components
-// import {terminateQueryAlertLocalSetting, terminateSessionAlertLocalSetting} from "src/redux/alerts";
-
 import { terminateQuery, terminateSession } from "src/api/terminateQueryApi";
 import { actions as sessionsActions } from "src/store/sessions";
 import { actions as terminateQueryActions } from "./terminateQuery.reducer";
 
 import { cockroach } from "@cockroachlabs/crdb-protobuf-client";
+import { XOR } from "src/util/xor";
+
 const CancelSessionRequest = cockroach.server.serverpb.CancelSessionRequest;
 const CancelQueryRequest = cockroach.server.serverpb.CancelQueryRequest;
-export type ICancelSessionRequest = cockroach.server.serverpb.CancelSessionRequest;
-export type ICancelQueryRequest = cockroach.server.serverpb.CancelQueryRequest;
+export type ICancelSessionRequest = cockroach.server.serverpb.ICancelSessionRequest;
+export type ICancelQueryRequest = cockroach.server.serverpb.ICancelQueryRequest;
+interface CancelSessionWithCallbacks {
+  req: ICancelSessionRequest;
+  onCancelSessionSuccess?: () => void;
+  onCancelSessionError?: () => void;
+}
+
+export type CancelSessionPayload = XOR<
+  CancelSessionWithCallbacks,
+  ICancelSessionRequest
+>;
```

Comment:
I don't think this is an easily understood pattern and it's making already complex code much harder to understand. Why not use a standard Union type `|` and do a switch when you need to decide which one to use?

Reply:
This approach needed to support 2 very different types of payloads. Standard union will show error something like "type  Foo do not correspond to type Foo | Bar, Bar is missing". We will have ability to remove all this workarounds as soon as move redux layer to CC. But it is needed for now.